wolfcrypt/test/test.c

```diff
@@ -49368,6 +49380,49 @@ static int myCryptoDevCb(int devIdArg, wc_CryptoInfo* info, void* ctx)
         info->hmac.hmac->devId = devIdArg;
     }
 #endif
+#if defined(WOLFSSL_CMAC) && !defined(NO_AES) && defined(WOLFSSL_AES_DIRECT)
+    else if (info->algo_type == WC_ALGO_TYPE_CMAC) {
+        if (info->cmac.cmac == NULL) {
+            return NOT_COMPILED_IN;
+        }
+
+        /* set devId to invalid so software is used */
+        info->cmac.cmac->devId = INVALID_DEVID;
+
+        if (info->cmac.in != NULL && info->cmac.out != NULL) {
+            ret = wc_AesCmacGenerate(info->cmac.out,
+                                     info->cmac.outSz,
+                                     info->cmac.in,
+                                     info->cmac.inSz,
+                                     info->cmac.key,
+                                     info->cmac.keySz);
+        }
+        else if (info->cmac.key != NULL) {
```

Comment:
These cases are not mutually exclusive but rather sequential. Change the non-oneshot logic to:
ret = 0;
if ((ret == 0) && (key !=NULL)) ret = InitCmac(...);
if ((ret == 0) && (in !=NULL)) ret = UpdateCmac(...);
if ((ret == 0) && (out !=NULL)) ret = FinalCmac(...);

Reply:
hmm I don't understand why they should be sequential rather than mutually exclusive. Isn't the point for there to be a 1-to-1 mapping between the WC API called and the case invoked here (e.g. the cryptoCb detecting whether it is being invoked from init/update/final)?  The user can only call one of  init, update, or final, and each of those functions passes a mutually exclusive set of args to the cryptocb (for example, there is no way for `wc_InitCmac` to pass a non-null `in` or `out` arg to the cryptoCb, and no way for `wc_CmacUpdate` to pass a non-null `out` arg to the cryptoCb) so why would we want to support "falling through" to these cases inside the callback?